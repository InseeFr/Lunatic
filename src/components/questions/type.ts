import { ReactNode } from 'react-markdown/lib/ast-to-react';

export type MessageType = {
	text?: ReactNode;
};
