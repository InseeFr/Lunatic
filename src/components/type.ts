import {
	LunaticComponentDefinition,
	LunaticError,
	LunaticExpression,
	LunaticState,
} from '../use-lunatic/type';
import { CSSProperties, FunctionComponent, ReactNode } from 'react';
import { SuggesterStatus } from '../use-lunatic/use-suggesters';
import useLunatic from '../use-lunatic';

export type VtlExpression = {
	value: string;
	type: 'VTL' | 'VTL|MD';
};

export type LunaticBaseProps<ValueType = unknown> = {
	id: string;
	handleChange: (
		response: { name: string },
		value: ValueType,
		args?: Record<string, unknown>
	) => void;
	errors?: { [id: string]: LunaticError[] };
	preferences?: LunaticState['preferences'];
	declarations?: {
		id: string;
		declarationType: string;
		position: string;
		label: ReactNode;
	}[];
	label?: ReactNode;
	disabled?: boolean;
	missing?: unknown;
	missingResponse?: { name: string; value?: unknown };
	management?: LunaticState['management'];
	description?: ReactNode;
	shortcut?: boolean;
	required?: boolean;
	value: null | ValueType;
	readOnly?: boolean;
	className?: string;
	style?: CSSProperties;
	iteration?: number;
	executeExpression: LunaticState['executeExpression'];
	features?: string[];
	componentType?: string;
};

export type SuggesterOption = {
	children?: string[];
	id?: string;
	description?: ReactNode;
	label?: ReactNode;
	value: string;
	niveau?: string;
	parent?: string;
	tokensMap?: Record<string, { count: number; fields: string[] }>;
};

type ComponentPropsByType = {
	InputNumber: LunaticBaseProps<number | null> & {
		min: number;
		max: number;
		decimals: number;
		unit?: string;
		response: { name: string };
	};
	Input: LunaticBaseProps<string> & {
		maxLength?: number;
		value: null | string;
		response: { name: string };
	};
	Sequence: Pick<
		LunaticBaseProps<string>,
		'id' | 'declarations' | 'label' | 'style'
	>;
	Subsequence: Pick<LunaticBaseProps<string>, 'id' | 'declarations' | 'label'>;
	Question: Pick<LunaticBaseProps<unknown>, 'label' | 'description'>;
	ComponentSet: LunaticBaseProps<unknown> & {
		components: LunaticComponentDefinition[];
		value: Record<string, unknown>;
		response: undefined;
	};
	RosterForLoop: LunaticBaseProps<unknown> & {
		lines: { min: number; max: number };
		iterations?: number;
		components: LunaticComponentDefinition[];
		executeExpression: LunaticState['executeExpression'];
		value: Record<string, unknown[]>;
		headers?: Array<{ label: ReactNode }>;
		paginatedLoop?: boolean;
	};
	Loop: LunaticBaseProps<unknown> & {
		lines: { min: number; max: number };
		iterations?: number;
		components: LunaticComponentDefinition[];
		executeExpression: LunaticState['executeExpression'];
		value: Record<string, unknown[]>;
		headers?: Array<{ label: ReactNode }>;
		paginatedLoop?: boolean;
	};
	Table: LunaticBaseProps<unknown> & {
		value: Record<string, unknown>;
		header: Array<{
			label: ReactNode;
			rowspan?: number;
			colspan?: number;
		}>;
		body: Array<Array<{ label: LunaticExpression }>>;
		executeExpression: LunaticState['executeExpression'];
		iteration: LunaticState['pager']['iteration'];
	};
	Datepicker: LunaticBaseProps<string | null> & {
		min?: string;
		max?: string;
		response: { name: string };
	};
	CheckboxGroup: LunaticBaseProps<Record<string, boolean | null>> & {
		responses: Array<{
			id: string;
			label: ReactNode;
			description?: ReactNode;
			response: { name: string };
		}>;
		handleChange: (
			response: { name: string },
			value: boolean,
			args?: Record<string, unknown>
		) => void;
	};
	CheckboxOne: LunaticBaseProps<string | null> & {
		options: Array<{
			description: ReactNode;
			label: ReactNode;
			value: string;
		}>;
		response: { name: string };
	};
	Switch: LunaticBaseProps<boolean> & {
		response: { name: string };
		statusLabel?: { true: string; false: string };
	};
	CheckboxBoolean: LunaticBaseProps<boolean> & {
		options: Array<{ description: ReactNode; label: ReactNode; value: string }>;
		response: { name: string };
	};
	Radio: LunaticBaseProps<string | null> & {
		options: Array<{ description: ReactNode; label: ReactNode; value: string }>;
		checkboxStyle?: boolean;
		response: { name: string };
	};
	Roundabout: LunaticBaseProps<string> & {
		iterations: number;
		goToPage: ReturnType<typeof useLunatic>['goToPage'];
		page: string;
		locked?: boolean;
		expressions: {
			unnecessary?: Array<boolean>;
			complete?: Array<boolean>;
			partial?: Array<boolean>;
			label?: Array<string>;
		};
	};
	Dropdown: LunaticBaseProps<string | null> & {
		options: Array<{ description: ReactNode; label: ReactNode; value: string }>;
		response: { name: string };
		writable?: boolean;
	};
	Textarea: LunaticBaseProps<string> & {
		cols?: number;
		placeHolder?: string;
		maxLength?: number;
		rows?: number;
		response: { name: string };
	};
	FilterDescription: Pick<LunaticBaseProps<string>, 'id' | 'label'>;
	Accordion: Pick<LunaticBaseProps<string>, 'id' | 'label' | 'description'> & { bgColor?: string };
	PairwiseLinks: Omit<LunaticBaseProps, 'value'> & {
		components: LunaticComponentDefinition[];
		features?: LunaticState['features'];
		executeExpression: LunaticState['executeExpression'];
		xAxisIterations: number;
		yAxisIterations: number;
		symLinks: Record<string, Record<string, string>>;
		value: Record<string, unknown[]>;
	};
	Suggester: LunaticBaseProps<string | null> & {
		storeName: string;
		getSuggesterStatus: (name: string) => {
			status: SuggesterStatus;
			timestamp: number;
		};
		optionRenderer: FunctionComponent<{
			option: SuggesterOption;
			placeholder?: string;
			search?: string;
		}>;
		labelRenderer: FunctionComponent<{
			option?: SuggesterOption;
			selected?: boolean;
			search?: string;
		}>;
		idbVersion?: string;
		focused: boolean;
		response: { name: string };
	};
	Summary: LunaticBaseProps<string | null> & {
		sections: Array<{
			responses?: Array<{ label: VtlExpression; value: VtlExpression }>;
			title?: VtlExpression;
			iterations?: number;
		}>
	};
};

export type LunaticComponentType = keyof ComponentPropsByType;

export type LunaticComponentProps<
	T extends LunaticComponentType = LunaticComponentType
> = ComponentPropsByType[T];
