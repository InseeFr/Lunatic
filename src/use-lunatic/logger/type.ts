type LoggerMessage =
	| {
			type: 'ERROR';
			error: Error;
	  }
	| {
			type: 'WARNING';
			message: string;
	  };

export type LunaticLogger = (msg: LoggerMessage) => void;
