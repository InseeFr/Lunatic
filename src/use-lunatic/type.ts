import type { ReactNode } from 'react';
import type {
	ComponentType,
	ControlType,
	LunaticSource,
	Variable,
} from './type-source';
import { SuggesterStatus } from './use-suggesters';
import type { LunaticVariablesStore } from './commons/variables/lunatic-variables-store';

export type LunaticComponentDefinition<
	T extends ComponentType['componentType'] = ComponentType['componentType']
> = ComponentType & { componentType: T };
export type LunaticControl = ControlType;

export type VTLBindings = { [variableName: string]: unknown };

export type LunaticData = Partial<
	Record<Exclude<VariableType, 'COLLECTED'>, Record<string, unknown>> & {
		COLLECTED: Record<string, LunaticCollectedValue>;
	}
>;

export type LunaticValues = {
	[variableName: string]: unknown;
};

export type LunaticError = Pick<
	ControlType,
	'id' | 'criticality' | 'typeOfControl'
> & {
	errorMessage: ReactNode;
};

export type VariableType = 'COLLECTED' | 'EXTERNAL' | 'CALCULATED';
export type ExpressionType = 'VTL' | 'VTL|MD';
export type LunaticExpression = {
	value: string;
	type: ExpressionType;
	bindingDependencies?: string[];
};
export type TODO = unknown; // Temporary type to mark types as unresolved

export type LunaticVariable = Variable;
export type LunaticCollectedValue = {
	COLLECTED: unknown;
	EDITED: unknown;
	FORCED: unknown;
	INPUTED: unknown;
	PREVIOUS: unknown;
};

export type LunaticOverviewItem = {
	id: string;
	page: string;
	type: string;
	evaluatedLabel: string;
	visible: boolean;
	reached: boolean;
	parent?: unknown;
	label: LunaticExpression;
	conditionFilter?: {
		bindingDependencies?: string[];
	};
	children: LunaticOverviewItem[];
};

// We need a mapped type to correlate type and variableType
export type LunaticStateVariable = {
	[key in LunaticVariable['variableType']]: {
		type: key;
		value: unknown;
		variable: LunaticVariable & { variableType: key };
		CalculatedLinked?: LunaticVariable[];
	};
}[LunaticVariable['variableType']];

export type LunaticState = {
	updatedAt: number;
	variables: LunaticVariablesStore;
	pages: {
		[key: number | string]:
			| {
					components: ComponentType[];
					isLoop: false;
					iterations?: undefined;
					loopDependencies?: undefined;
					subPages?: undefined;
			  }
			| {
					components: ComponentType[];
					isLoop: true;
					iterations: ExpressionType;
					// Variables affecting this loop
					loopDependencies: string[];
					// List of child pages (ex: ['20.1', '20.2']
					subPages: string[];
			  };
	};
	isInLoop: boolean;
	isFirstPage: boolean;
	isLastPage: boolean;
	features: ['VTL'] | ['VTL', 'MD'];
	preferences: ['COLLECTED'];
	savingType: 'COLLECTED';
	// Map of variable associated with the expression used to repopulate it
	cleaning: LunaticSource['cleaning'];
	// Map of variable with the missing variable names associated with it ex: {ADR_COLL: ['ADR_COLL_MISSING']}
	missingBlock: { [variable: string]: string[] };
	// Map of variable with the missing variable names associated with it ex: {ADR_COLL: ['ADR_COLL_MISSING']}
	resizing: {
		[variable: string]: {
			// VTL expression
			size: string;
			// List of variables that will need resizing
			variables: string[];
			sizeForLinksVariables?: unknown;
			linksVariables?: string[];
		};
	};
	overview: LunaticOverviewItem[];
	pager: {
		lastReachedPage?: string;
		maxPage: string;
		nbSubPages?: number;
		page: string;
		subPage?: number;
		// Iteration index (starting at 0)
		iteration?: number;
		nbIterations?: number;
		shallowIteration?: number;
		linksIterations?: number[];
	};
	// TODO : Explain this
	waiting: boolean;
	// Errors for the form
	errors?: { [page: string]: { [id: string]: LunaticError[] } };
	// Contains the errors for the current page / iteration
	currentErrors?: { [id: string]: LunaticError[] };
	// Errors
	modalErrors?: Record<string, LunaticError[]>;
	// Handler to call when updating a value
	handleChange: (
		response: { name: string },
		value: any,
		args?: Record<string, unknown>
	) => void;
	// Run and expression using the value from the state
	executeExpression: <T extends unknown = unknown>(
		expression: unknown,
		args?: {
			iteration?: number;
			// @deprecated
			bindingDependencies?: string[];
			deps?: string[];
		}
	) => T;
	// Update the value collected for the variable
	updateBindings: (variableName: string, value: unknown) => unknown;
	// Enable controls for data (form validation)
	activeControls: boolean;
	// enable shortcut on radio/checkbox/missing buttons
	shortcut?: boolean;
	// TODO : Explain this
	management?: boolean;
	goToPage: (page: {
		page: string;
		iteration?: number;
		nbIterations?: number;
		subPage?: number;
	}) => void;
	// Enable components to independently navigate next/previous
	goNextPage: () => void;
	goPreviousPage: () => void;
	getSuggesterStatus: (name: string) => {
		status: SuggesterStatus;
		timestamp: number;
	};
};
