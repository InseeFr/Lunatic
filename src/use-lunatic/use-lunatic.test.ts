import { act, renderHook } from '@testing-library/react-hooks';
import { describe, expect, it } from 'vitest';
import useLunatic from './use-lunatic';

import sourceWithoutHierarchy from '../stories/overview/source.json';
import sourceLogement from '../stories/questionnaires/logement/source.json';
import sourceSimpsons from '../stories/questionnaires/simpsons/source.json';
import { LunaticData } from './type';

const dataFromObject = (o: Record<string, unknown>): LunaticData => {
	return {
		EXTERNAL: {},
		COLLECTED: Object.keys(o).reduce(
			(acc, k) => ({
				...acc,
				[k]: {
					EDITED: null,
					FORCED: null,
					INPUTED: null,
					PREVIOUS: null,
					COLLECTED: o[k],
				},
			}),
			{}
		),
		CALCULATED: {},
	};
};

describe('use-lunatic()', () => {
	const defaultParams = [
		sourceSimpsons as any,
		dataFromObject({}),
		{},
	] as const;

	it('should initialize correcly', () => {
		const { result } = renderHook(() => useLunatic(...defaultParams));
		expect(result.current.pager.page).toBe('1');
		expect(result.current.pager.lastReachedPage).toBe('1');
		expect(result.current.pager.maxPage).toBe('39');
	});
	it('should go to the next page correcly', () => {
		const { result } = renderHook(() => useLunatic(...defaultParams));
		expect(result.current.pager.page).toBe('1');
		expect(result.current.pager.lastReachedPage).toBe('1');
		act(() => {
			result.current.goNextPage();
		});
		expect(result.current.pager.page).toBe('2');
		expect(result.current.pager.lastReachedPage).toBe('2');
	});
	it('should jump to a specific page', () => {
		const params = [
			sourceSimpsons as any,
			dataFromObject({
				COMMENT: 'Hello world',
				READY: true,
				NAME_CHAR: ['a', 'b'],
			}),
			{
				initialPage: '35.1#1',
			},
		] as const;
		const { result } = renderHook(() => useLunatic(...params));
		const components = result.current.getComponents();
		expect(result.current.pager.lastReachedPage).toBe('35.1#1');
		expect(result.current.pager.iteration).toBe(0);
		expect(result.current.pager.subPage).toBe(0);
		expect(result.current.pager.page).toBe('35');
		expect(components[0].id).toBe('kiq5xw5p');
	});

	describe('Provider', () => {
		it('should not generate a new Provider every render', () => {
			const { result } = renderHook(() => useLunatic(...defaultParams));
			const oldProvider = result.current.Provider;
			act(() => {
				result.current.goNextPage();
			});
			expect(result.current.Provider).toBe(oldProvider);
		});
	});

	describe('overview', function () {
		const lunaticConfigurationWithoutOverview = {
			management: false,
			activeControls: false,
			initialPage: '1',
			getStoreInfo: () => {},
			missing: false,
			shortcut: false,
			activeGoNextForMissing: false,
			showOverview: false,
			filterDescription: true,
		};

		const lunaticConfiguration = {
			...lunaticConfigurationWithoutOverview,
			withOverview: true,
		};

		const advancedQestionnaireData = dataFromObject({ CADR: '1' });

		it('should not calculate overview by default', function () {
			const { result } = renderHook(() =>
				useLunatic(
					sourceLogement as any,
					undefined,
					lunaticConfigurationWithoutOverview
				)
			);
			const overview = result.current.overview;
			expect(overview).toHaveLength(0);
		});
		it('should make the first sequence visible', function () {
			const { result } = renderHook(() =>
				useLunatic(sourceLogement as any, undefined, lunaticConfiguration)
			);
			const overview = result.current.overview;
			expect(overview).toHaveLength(11);
			expect(overview[0].reached).toEqual(true);
			expect(overview[0].visible).toEqual(true);
			expect(overview[1].reached).toEqual(false);
			expect(overview[1].visible).toEqual(true);
		});
		it('should update the breadcrumb on page change', function () {
			const { result } = renderHook(() =>
				useLunatic(sourceLogement as any, undefined, lunaticConfiguration)
			);
			// '8' page should not be visible
			expect(result.current.overview[0].children[0].visible).toBe(false);
			// Simulate a change of value in the form
			act(() => {
				const c = result.current.getComponents();
				c[0].handleChange({ name: 'CADR' }, '3', {});
				result.current.goNextPage();
			});
			// Page '8' should now be visible
			expect(result.current.overview[0].children[0].visible).toBe(true);
		});
		it('should be empty when no hierarchy', function () {
			const { result } = renderHook(() =>
				useLunatic(
					sourceWithoutHierarchy as any,
					undefined,
					lunaticConfiguration
				)
			);
			expect(result.current.overview).toHaveLength(0);
		});

		describe('with initial data', function () {
			it('should make second sequence visible', function () {
				const { result } = renderHook(() =>
					useLunatic(sourceLogement as any, advancedQestionnaireData, {
						...lunaticConfiguration,
						// hack on initialPage : useLunatic SHOULD find lastReachedPage from COLLECTED data
						initialPage: '1',
						// use lastReachedPage saved by orchestrator
						lastReachedPage: '16',
					})
				);
				const overview = result.current.overview;
				expect(overview).toHaveLength(11);
				expect(overview[0].reached).toEqual(true);
				expect(overview[0].visible).toEqual(true);
				expect(overview[1].reached).toEqual(true);
				expect(overview[1].visible).toEqual(true);
			});
		});
	});
});
