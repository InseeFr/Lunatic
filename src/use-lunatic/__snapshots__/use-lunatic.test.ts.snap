// Vitest Snapshot v1

exports[`use-lunatic() > overview > with loop > should handle lastReachedPage 1`] = `
[
  {
    "children": [
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "lujqeci5",
        "label": <MDLabel
          expression="Votre santé"
        />,
        "page": "2",
        "reached": true,
        "type": "Subsequence",
      },
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "lujqbyzl",
        "label": <MDLabel
          expression="Votre logement"
        />,
        "page": "3",
        "reached": true,
        "type": "Subsequence",
      },
    ],
    "current": true,
    "description": undefined,
    "id": "lujqfpva",
    "label": <MDLabel
      expression="I - Séquence 1"
    />,
    "page": "1",
    "reached": true,
    "type": "Sequence",
  },
  {
    "children": [
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "lulbmyhr",
        "label": <MDLabel
          expression="Commentaire sur composition du logement"
        />,
        "page": "8",
        "reached": true,
        "type": "Subsequence",
      },
    ],
    "current": false,
    "description": undefined,
    "id": "lujqrqmp",
    "label": <MDLabel
      expression="II - Habitants du logement"
    />,
    "page": "5",
    "reached": true,
    "type": "Sequence",
  },
  {
    "children": [
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "lujykwaz",
        "label": <MDLabel
          expression="Caractéristiques de Quentin"
        />,
        "page": "10.1#1",
        "reached": true,
        "type": "Subsequence",
      },
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "lujykwaz",
        "label": <MDLabel
          expression="Caractéristiques de Luna"
        />,
        "page": "10.1#2",
        "reached": true,
        "type": "Subsequence",
      },
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "lujykwaz",
        "label": <MDLabel
          expression="Caractéristiques de Paul"
        />,
        "page": "10.1#3",
        "reached": true,
        "type": "Subsequence",
      },
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "lujyik5q",
        "label": <MDLabel
          expression="Autres caractéristiques de Luna : filtrée pour le premier individu"
        />,
        "page": "11.1#2",
        "reached": true,
        "type": "Subsequence",
      },
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "luk0swcz",
        "label": <MDLabel
          expression="Encore d'autres caractéristiques de Luna : filtrée pour le premier individu"
        />,
        "page": "11.2#2",
        "reached": true,
        "type": "Subsequence",
      },
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "lujyik5q",
        "label": <MDLabel
          expression="Autres caractéristiques de Paul : filtrée pour le premier individu"
        />,
        "page": "11.1#3",
        "reached": false,
        "type": "Subsequence",
      },
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "luk0swcz",
        "label": <MDLabel
          expression="Encore d'autres caractéristiques de Paul : filtrée pour le premier individu"
        />,
        "page": "11.2#3",
        "reached": false,
        "type": "Subsequence",
      },
    ],
    "current": false,
    "description": undefined,
    "id": "lujyi4pe",
    "label": <MDLabel
      expression="III - Détail des individus"
    />,
    "page": "9",
    "reached": true,
    "type": "Sequence",
  },
  {
    "children": [
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "lumfc98o",
        "label": <MDLabel
          expression="Belle sous-séquence pour Quentin"
        />,
        "page": "12.2#1",
        "reached": false,
        "type": "Subsequence",
      },
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "lumfe3bj",
        "label": <MDLabel
          expression="Autre belle sous-seq"
        />,
        "page": "12.3#1",
        "reached": false,
        "type": "Subsequence",
      },
    ],
    "current": false,
    "description": undefined,
    "id": "luk1ojt5",
    "label": <MDLabel
      expression="IV - Belle séquence pour Quentin"
    />,
    "page": "12.1#1",
    "reached": false,
    "type": "Sequence",
  },
  {
    "children": [],
    "current": false,
    "description": undefined,
    "id": "lulbelgr",
    "label": <MDLabel
      expression="V - Autre séquence pour Quentin"
    />,
    "page": "12.4#1",
    "reached": false,
    "type": "Sequence",
  },
  {
    "children": [
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "lumfc98o",
        "label": <MDLabel
          expression="Belle sous-séquence pour Luna"
        />,
        "page": "12.2#2",
        "reached": false,
        "type": "Subsequence",
      },
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "lumfe3bj",
        "label": <MDLabel
          expression="Autre belle sous-seq"
        />,
        "page": "12.3#2",
        "reached": false,
        "type": "Subsequence",
      },
    ],
    "current": false,
    "description": undefined,
    "id": "luk1ojt5",
    "label": <MDLabel
      expression="IV - Belle séquence pour Luna"
    />,
    "page": "12.1#2",
    "reached": false,
    "type": "Sequence",
  },
  {
    "children": [],
    "current": false,
    "description": undefined,
    "id": "lulbelgr",
    "label": <MDLabel
      expression="V - Autre séquence pour Luna"
    />,
    "page": "12.4#2",
    "reached": false,
    "type": "Sequence",
  },
  {
    "children": [
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "lumfc98o",
        "label": <MDLabel
          expression="Belle sous-séquence pour Paul"
        />,
        "page": "12.2#3",
        "reached": false,
        "type": "Subsequence",
      },
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "lumfe3bj",
        "label": <MDLabel
          expression="Autre belle sous-seq"
        />,
        "page": "12.3#3",
        "reached": false,
        "type": "Subsequence",
      },
    ],
    "current": false,
    "description": undefined,
    "id": "luk1ojt5",
    "label": <MDLabel
      expression="IV - Belle séquence pour Paul"
    />,
    "page": "12.1#3",
    "reached": false,
    "type": "Sequence",
  },
  {
    "children": [],
    "current": false,
    "description": undefined,
    "id": "lulbelgr",
    "label": <MDLabel
      expression="V - Autre séquence pour Paul"
    />,
    "page": "12.4#3",
    "reached": false,
    "type": "Sequence",
  },
  {
    "children": [],
    "current": false,
    "description": undefined,
    "id": "COMMENT-SEQ",
    "label": <MDLabel
      expression="Commentaire"
    />,
    "page": "13",
    "reached": false,
    "type": "Sequence",
  },
]
`;

exports[`use-lunatic() > overview > with loop > should work with loop 1`] = `
[
  {
    "children": [
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "lujqeci5",
        "label": <MDLabel
          expression="Votre santé"
        />,
        "page": "2",
        "reached": false,
        "type": "Subsequence",
      },
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "lujqbyzl",
        "label": <MDLabel
          expression="Votre logement"
        />,
        "page": "3",
        "reached": false,
        "type": "Subsequence",
      },
    ],
    "current": true,
    "description": undefined,
    "id": "lujqfpva",
    "label": <MDLabel
      expression="I - Séquence 1"
    />,
    "page": "1",
    "reached": true,
    "type": "Sequence",
  },
  {
    "children": [
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "lulbmyhr",
        "label": <MDLabel
          expression="Commentaire sur composition du logement"
        />,
        "page": "8",
        "reached": false,
        "type": "Subsequence",
      },
    ],
    "current": false,
    "description": undefined,
    "id": "lujqrqmp",
    "label": <MDLabel
      expression="II - Habitants du logement"
    />,
    "page": "5",
    "reached": false,
    "type": "Sequence",
  },
  {
    "children": [
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "lujykwaz",
        "label": <MDLabel
          expression="Caractéristiques de Quentin"
        />,
        "page": "10.1#1",
        "reached": false,
        "type": "Subsequence",
      },
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "lujykwaz",
        "label": <MDLabel
          expression="Caractéristiques de Luna"
        />,
        "page": "10.1#2",
        "reached": false,
        "type": "Subsequence",
      },
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "lujykwaz",
        "label": <MDLabel
          expression="Caractéristiques de Paul"
        />,
        "page": "10.1#3",
        "reached": false,
        "type": "Subsequence",
      },
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "lujyik5q",
        "label": <MDLabel
          expression="Autres caractéristiques de Luna : filtrée pour le premier individu"
        />,
        "page": "11.1#2",
        "reached": false,
        "type": "Subsequence",
      },
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "luk0swcz",
        "label": <MDLabel
          expression="Encore d'autres caractéristiques de Luna : filtrée pour le premier individu"
        />,
        "page": "11.2#2",
        "reached": false,
        "type": "Subsequence",
      },
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "lujyik5q",
        "label": <MDLabel
          expression="Autres caractéristiques de Paul : filtrée pour le premier individu"
        />,
        "page": "11.1#3",
        "reached": false,
        "type": "Subsequence",
      },
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "luk0swcz",
        "label": <MDLabel
          expression="Encore d'autres caractéristiques de Paul : filtrée pour le premier individu"
        />,
        "page": "11.2#3",
        "reached": false,
        "type": "Subsequence",
      },
    ],
    "current": false,
    "description": undefined,
    "id": "lujyi4pe",
    "label": <MDLabel
      expression="III - Détail des individus"
    />,
    "page": "9",
    "reached": false,
    "type": "Sequence",
  },
  {
    "children": [
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "lumfc98o",
        "label": <MDLabel
          expression="Belle sous-séquence pour Quentin"
        />,
        "page": "12.2#1",
        "reached": false,
        "type": "Subsequence",
      },
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "lumfe3bj",
        "label": <MDLabel
          expression="Autre belle sous-seq"
        />,
        "page": "12.3#1",
        "reached": false,
        "type": "Subsequence",
      },
    ],
    "current": false,
    "description": undefined,
    "id": "luk1ojt5",
    "label": <MDLabel
      expression="IV - Belle séquence pour Quentin"
    />,
    "page": "12.1#1",
    "reached": false,
    "type": "Sequence",
  },
  {
    "children": [],
    "current": false,
    "description": undefined,
    "id": "lulbelgr",
    "label": <MDLabel
      expression="V - Autre séquence pour Quentin"
    />,
    "page": "12.4#1",
    "reached": false,
    "type": "Sequence",
  },
  {
    "children": [
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "lumfc98o",
        "label": <MDLabel
          expression="Belle sous-séquence pour Luna"
        />,
        "page": "12.2#2",
        "reached": false,
        "type": "Subsequence",
      },
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "lumfe3bj",
        "label": <MDLabel
          expression="Autre belle sous-seq"
        />,
        "page": "12.3#2",
        "reached": false,
        "type": "Subsequence",
      },
    ],
    "current": false,
    "description": undefined,
    "id": "luk1ojt5",
    "label": <MDLabel
      expression="IV - Belle séquence pour Luna"
    />,
    "page": "12.1#2",
    "reached": false,
    "type": "Sequence",
  },
  {
    "children": [],
    "current": false,
    "description": undefined,
    "id": "lulbelgr",
    "label": <MDLabel
      expression="V - Autre séquence pour Luna"
    />,
    "page": "12.4#2",
    "reached": false,
    "type": "Sequence",
  },
  {
    "children": [
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "lumfc98o",
        "label": <MDLabel
          expression="Belle sous-séquence pour Paul"
        />,
        "page": "12.2#3",
        "reached": false,
        "type": "Subsequence",
      },
      {
        "children": [],
        "current": false,
        "description": undefined,
        "id": "lumfe3bj",
        "label": <MDLabel
          expression="Autre belle sous-seq"
        />,
        "page": "12.3#3",
        "reached": false,
        "type": "Subsequence",
      },
    ],
    "current": false,
    "description": undefined,
    "id": "luk1ojt5",
    "label": <MDLabel
      expression="IV - Belle séquence pour Paul"
    />,
    "page": "12.1#3",
    "reached": false,
    "type": "Sequence",
  },
  {
    "children": [],
    "current": false,
    "description": undefined,
    "id": "lulbelgr",
    "label": <MDLabel
      expression="V - Autre séquence pour Paul"
    />,
    "page": "12.4#3",
    "reached": false,
    "type": "Sequence",
  },
  {
    "children": [],
    "current": false,
    "description": undefined,
    "id": "COMMENT-SEQ",
    "label": <MDLabel
      expression="Commentaire"
    />,
    "page": "13",
    "reached": false,
    "type": "Sequence",
  },
]
`;
